decidim/decidim
# Create actionable notifications

#### :tophat: What? Why?

This PR adds an actionable notifications framework. 
It uses two optional methods in the event class attached to a notification model. If that Event returns a cell to handle the actions it will be rendered.

The actionable zone uses Cells allowing for custom types of actions in it. By default we add a ButtonsCell for actions involving one or more buttons with some HTTP action in it.

ButtonsCell Actionable workflow:

1. Extract information from the `action_data` method (see code for details) from the Event attached into the notification. This includes the URL where to post/get/patch/delete when clicking for every button needed (typically "Accept"/"Reject").
2. Extract the type of cell used for rendering the action. If the cell does not exist, it just ignores it.
3. Render the buttons for ajax remote operation.
4. After some action is performed (after click and waiting for server response), it removes the buttons and replaces them by the answer returned by the url in the action. If no message is returned, it just removes the buttons.
5. In the action had some problems, a message is rendered instead of the buttons (but this is not permanent)

Example of the JSON for the `action_cell` and `action_data` methods for rendering 2 buttons for accepting/rejecting group membership. This would render the action using the ButtonsCell class:

```ruby
  # decidim-core/app/events/decidim/invited_to_group_event.rb
  def membership_id
    extra["membership_id"]
  end

  def invitation
    @invitation ||= UserGroupMembership.find_by(user:, id: membership_id, role: "invited")
  end

  def action_cell
    "decidim/notification_actions/buttons" if invitation
  end

  def action_data
    [
      {
        url: url_helpers.group_invite_path(user_group_nickname, membership_id, format: :json),
        icon: "check-line",
        method: "patch",
        i18n_label: "decidim.group_invites.accept_invitation"
      },
      {
        url: url_helpers.group_invite_path(user_group_nickname, membership_id, format: :json),
        icon: "close-circle-line",
        method: "delete",
        i18n_label: "decidim.group_invites.reject_invitation"
      }
    ]
  end
```

As this introduces a new internal API, I've added documentation too.

Tasks

- [x] Actionable framework for notifications
- [x] Basic action cells (button, callout)
- [x] Specs for standard URL calls (handles flash messages)
- [x] Specs for JSON calls (handles response message)
- [x] Add example for (group accept/reject membership)
- [x] ASCIIDoc documentation

#### :pushpin: Related Issues
*Link your PR to an issue*
- Related to #12362 
- Fixes #?

#### Testing

1. Open two sessions for different Decidim users
2. In one session, go to a user group and invite the other user as a member of the group
3. Go to the other session, navigate to /notifications
4. See the buttons to accept/reject the group invitation. Proceed as you wish
5. Reload the page, no buttons will appear anymore

### :camera: Screenshots

![](https://github.com/decidim/decidim/assets/1401520/7abb8a84-89c5-4be7-b9d6-c4bf9df34bc9)
![](https://github.com/decidim/decidim/assets/1401520/b04ccef5-6303-4bc8-adef-f2a24c0d99c7)
![](https://github.com/decidim/decidim/assets/1401520/e0104be3-32c1-4af1-9dcc-3d0f9f7b636b)

:hearts: Thank you!
